mitodl/ocw-studio
# Remove extra parameter from < resource > shortcodes

#### Pre-Flight checklist


#### What are the relevant tickets? #1123 

#### What's this PR do?
When converting some images from  markdown `![alt text](file url)` to resource shortcodes, I erroneously included the alt text as an argument to the short code: `< resource uuid "alt text here" >`. This was *wrong:* resource shortcode only takes 1 uuid argument, the alt text etc is read from metadata.

This PR adds a rule to remove extra shortcode arguments.

This PR also...
- introduces the use of [pyparsing](https://pyparsing-docs.readthedocs.io/en/latest/pyparsing.html) as in markdown_cleanup rules
- removes a few old rules we seem unlikely to use again
- switches from batch updates to one-by-one updates

    This is less efficient but... it's not like we are updating *that* many records. And now we don't have to manually track sync_state changes. 

### How should this be manually tested?

1. Ensure you have some bad resource shortcodes... If you do not have recent rc/prod data, run `docker-compose run --rm web python manage.py markdown_cleanup rootrelative_urls -o rootrelative_urls.csv --commit` **on master branch**. This will make some new resource shortcodes that have too many arguments. (😱). Add `--skip-sync` if you want to skip the sync to github.
2. Checkout this branch.
3. Run `docker-compose run --rm web python manage.py markdown_cleanup remove_extra_resource_args -o remove_extra_args.csv --csv-only-changes` to generate a CSV of changes that will be made. This takes about 5 minutes.

5. Verify nothing has changed. (`dac7dbe9-bb2e-33c6-6027-a326888523c0` is a good page to inspect)
6. `docker-compose run --rm web python manage.py markdown_cleanup remove_extra_resource_args -o remove_extra_args.csv --commit`
7. Verify changes have been made (`dac7dbe9-bb2e-33c6-6027-a326888523c0` again)

You could run `manage markdown_cleanup shortcode_logging -o shortcodes.csv`, too. But it takes a while and does not make any changes, only logs shortcode usage.

#### Any background context you want to provide?

Why Pyparsing? Because:
- it's very flexible
- its `parse_action` and `transform_string` methods provide a mechanism similar to regex substitution, so making substitutions on part of the markdown is easy(ish) 
- and I've used it before so I knew it would work. E.g., EdX's formula parser is written with PyParsing (https://github.com/openedx/openedx-calc/blob/master/calc/calc.py)

Downsides... 
- significantly slower than regexp
- no special utilities for helping with markdown

We are not too tied to pyparsing. Most of `markdown_cleanup` is independent of it. https://readthedocs.org/projects/marko/ is maybe worth looking into more, too.
